File: pylith/faults/EqKinSrc.py
#!/usr/bin/env python
#
# ----------------------------------------------------------------------
#
# Brad T. Aagaard, U.S. Geological Survey
# Charles A. Williams, GNS Science
# Matthew G. Knepley, University of Chicago
#
# This code was developed as part of the Computational Infrastructure
# for Geodynamics (http://geodynamics.org).
#
# Copyright (c) 2010-2015 University of California, Davis
#
# See COPYING for license information.
#
# ----------------------------------------------------------------------
#

## @file pylith/faults/EqKinSrc.py
##

## @brief Python object for managing parameters for a kinematic
## earthquake sources.
##
## EqKinSrc is responsible for providing the value of slip at time t
## over a fault surface.
##
## Factory: eq_kinematic_src

from pylith.utils.PetscComponent import PetscComponent
from faults import EqKinSrc as ModuleEqKinSrc

# EqKinSrc class
class EqKinSrc(PetscComponent, ModuleEqKinSrc):
  """
  Python object for managing parameters for a kinematic earthquake sources.

  Inventory

  \b Properties
  @li \b origin_time Origin time for earthquake rupture.
  
  \b Facilities
  @li \b slip_function Slip time history function.

  Factory: eq_kinematic_src
  """

  # INVENTORY //////////////////////////////////////////////////////////

  import pyre.inventory
  
  from pyre.units.time import second
  originTime = pyre.inventory.dimensional("origin_time", default=0.0*second)
  originTime.meta['tip'] = "Origin time for earthquake rupture."
  
  from StepSlipFn import StepSlipFn
  slipfn = pyre.inventory.facility("slip_function", family="slip_time_fn",
                                   factory=StepSlipFn)
  slipfn.meta['tip'] = "Slip time history function."
  

  # PUBLIC METHODS /////////////////////////////////////////////////////

  def __init__(self, name="eqkinsrc"):
    """
    Constructor.
    """
    PetscComponent.__init__(self, name, facility="eqkinsrc")
    self._createModuleObj()
    self._loggingPrefix = "EqKi "
    return


  def preinitialize(self):
    """
    Do pre-initialization setup.
    """
    self._setupLogging()
    self.slipfn.preinitialize()
    return
  

  def verifyConfiguration(self):
    """
    Verify compatibility of configuration.
    """
    logEvent = "%sverify" % self._loggingPrefix
    self._eventLogger.eventBegin(logEvent)

    self.slipfn.verifyConfiguration()

    self._eventLogger.eventEnd(logEvent)
    return


  def initialize(self):
    """
    Initialize.
    """
    logEvent = "%sinit" % self._loggingPrefix
    self._eventLogger.eventBegin(logEvent)

    self.slipfn.initialize()

    self._eventLogger.eventEnd(logEvent)
    return


  def finalize(self):
    """
    Cleanup.
    """
    self.slipfn.finalize()
    return
  

  # PRIVATE METHODS ////////////////////////////////////////////////////

  def _configure(self):
    """
    Setup members using inventory.
    """
    PetscComponent._configure(self)
    ModuleEqKinSrc.originTime(self, self.inventory.originTime.value)
    ModuleEqKinSrc.slipfn(self, self.inventory.slipfn)
    return

  
  def _createModuleObj(self):
    """
    Create handle to corresponding C++ object.
    """
    ModuleEqKinSrc.__init__(self)
    return
  

  def _setupLogging(self):
    """
    Setup event logging.
    """
    if not "_loggingPrefix" in dir(self):
      self._loggingPrefix = ""

    from pylith.utils.EventLogger import EventLogger
    logger = EventLogger()
    logger.className("Kinematic Earthquake Source")
    logger.initialize()

    events = ["verify",
              "init"]
    for event in events:
      logger.registerEvent("%s%s" % (self._loggingPrefix, event))

    self._eventLogger = logger
    return
  

# FACTORIES ////////////////////////////////////////////////////////////

def eq_kinematic_src():
  """
  Factory associated with EqKinSrc.
  """
  return EqKinSrc()


# End of file 

File: libsrc/pylith/faults/EqKinSrc.cc
// -*- C++ -*-
//
// ----------------------------------------------------------------------
//
// Brad T. Aagaard, U.S. Geological Survey
// Charles A. Williams, GNS Science
// Matthew G. Knepley, University of Chicago
//
// This code was developed as part of the Computational Infrastructure
// for Geodynamics (http://geodynamics.org).
//
// Copyright (c) 2010-2015 University of California, Davis
//
// See COPYING for license information.
//
// ----------------------------------------------------------------------
//

#include <portinfo>

#include "EqKinSrc.hh" // implementation of object methods

#include "SlipTimeFn.hh" // USES SlipTimeFn

#include "pylith/utils/error.h" // USES PYLITH_METHOD_BEGIN

#include "spatialdata/units/Nondimensional.hh" // USES Nondimensional

#include <cassert> // USES assert()

// ----------------------------------------------------------------------
// Default constructor.
pylith::faults::EqKinSrc::EqKinSrc(void) :
  _originTime(0.0),
  _slipfn(0)
{ // constructor
} // constructor

// ----------------------------------------------------------------------
// Destructor.
pylith::faults::EqKinSrc::~EqKinSrc(void)
{ // destructor
  deallocate();
} // destructor

// ----------------------------------------------------------------------
// Deallocate PETSc and local data structures.
void 
pylith::faults::EqKinSrc::deallocate(void)
{ // deallocate
  _slipfn = 0; // :TODO: Use shared pointer.
} // deallocate
  
// ----------------------------------------------------------------------
// Set origin time for earthquake source.
void
pylith::faults::EqKinSrc::originTime(const PylithScalar value)
{ // originTime
  _originTime = value;
} // originTime

// ----------------------------------------------------------------------
// Get origin time for earthquake source.
PylithScalar
pylith::faults::EqKinSrc::originTime(void) const
{ // originTime
  return _originTime;
} // originTime

// ----------------------------------------------------------------------
// Set slip time function.
void
pylith::faults::EqKinSrc::slipfn(SlipTimeFn* slipfn)
{ // slipfn
  _slipfn = slipfn; // :TODO: Use shared pointer.
} // slipfn

// ----------------------------------------------------------------------
// Initialize slip time function.
void
pylith::faults::EqKinSrc::initialize(const topology::Mesh& faultMesh,
				     const spatialdata::units::Nondimensional& normalizer)
{ // initialize
  PYLITH_METHOD_BEGIN;

  // :TODO: Normalize slip time in Python?
  normalizer.nondimensionalize(&_originTime, 1, normalizer.timeScale());
  assert(_slipfn);
  _slipfn->initialize(faultMesh, normalizer, _originTime);

  PYLITH_METHOD_END;
} // initialize

// ----------------------------------------------------------------------
// Get slip on fault surface at time t.
void
pylith::faults::EqKinSrc::slip(topology::Field* const slipField,
			       const PylithScalar t)
{ // slip
  PYLITH_METHOD_BEGIN;

  assert(_slipfn);
  _slipfn->slip(slipField, t);

  PYLITH_METHOD_END;
} // slip

// ----------------------------------------------------------------------
// Get final slip.
const pylith::topology::Field&
pylith::faults::EqKinSrc::finalSlip(void) const
{ // finalSlip
  PYLITH_METHOD_BEGIN;

  assert(_slipfn);
  PYLITH_METHOD_RETURN(_slipfn->finalSlip());
} // finalSlip

// ----------------------------------------------------------------------
// Get time when slip begins at each point.
const pylith::topology::Field&
pylith::faults::EqKinSrc::slipTime(void) const
{ // slipTime
  PYLITH_METHOD_BEGIN;

  assert(_slipfn);
  PYLITH_METHOD_RETURN(_slipfn->slipTime());
} // slipTime


// End of file 
